### PR TITLE
Adds GDAS GEMPAK jobs to rocoto task mesh

### LINKED ISSUE #1219: Development workflow lacks GDAS gempak job
**Description**
The development (rocoto) workflow does not include a gdas gempak jobs in the task mesh, so `JGDAS_ATMOS_GEMPAK` and `JGDAS_ATMOS_GEMPAK_META_NCDC` never get exercised.

**Requirements**
Add a GDAS version of the gempak tasks when `DO_GEMPAK` is `YES`.

**Acceptance Criteria (Definition of Done)**
A gfsgempak and a gdasgempak job when gempak jobs are desired.

### PR DESCRIPTION
**Pull request by HenryRWinterbottom**
# Description

This PR addresses issue #1219. The following was accomplished:

- Rocoto task `jobs/rocoto/gempak.sh` was updated to support `JGDAS_ATMOS_GEMPAK` and `JGFS_ATMOS_GEMPAK`;
- Rocoto task `jobs/rocoto/gempakmeta.sh` was added to support `JGFS_ATMOS_GEMPAK_META`;
- Rocoto task `jobs/rocoto/gempakmetancdc.sh` was added to support `JGDAS_ATMOS_GEMPAK_META_NCDC`;
- Rocoto task `jobs/rocoto/gempakncdcupapgif.sh` was added to support `JGFS_ATMOS_GEMPAK_NCDC_UPAPGIF`;
- Rocoto task `jobs/rocoto/gempakpgrb2spec.sh` was added to support `JGFS_ATMOS_GEMPAK_PGRB2_SPEC`;
- The `npoess` switch is removed and the NPOESS task is renamed to `npoess_pgrb2_0p5deg` as per @KateFriedman-NOAA's requests (please see notes below).

  Resolves #1219 
  References #1222 #2073

**NOTE** This is a replicated PR open to resolve developer mishap with Github interface.

**NOTE** @KateFriedman-NOAA The `npoess` switch has been removed and the `npoess_pgrb2_0p5deg` task is now grouped beneath the `gempak` switches. This assumes that the `npoess_pgrb2_0p5deg` and all `gempak`-type tasks will use the same resources. However, the resource requirements (from `parm/config/gfs/config.resources` are different:

```
elif [[ ${step} = "npoess" ]]; then

    export wtime_npoess="03:30:00"
    export npe_npoess=1
    export npe_node_npoess=1
    export nth_npoess=1
    export memory_npoess="3GB"

elif [[ ${step} = "gempak" ]]; then

    export wtime_gempak="03:00:00"
    export npe_gempak=2
    export npe_gempak_gfs=28
    export npe_node_gempak=2
    export npe_node_gempak_gfs=28
    export nth_gempak=1
    export memory_gempak="4GB"
    export memory_gempak_gfs="2GB"
```

I am not sure how this is to be resolved and as a result I have maintained the `NPOESS` switch [here](https://github.com/HenryWinterbottom-NOAA/global-workflow/blob/feature/gfsv17_issue_1219.fix.002/parm/config/gfs/config.base.emc.dyn#L60.) until there is a resolution.

# Type of change

- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

The Rocoto XML configuration has been executed as follows.

```
user@host$: cd /scratch1/NCEPDEV/da/Henry.Winterbottom/trunk/global-workflow.develop.20231117/workflow
user@host$: ./setup_xml.py /scratch1/NCEPDEV/da/Henry.Winterbottom/work/GLOBAL_WORKFLOW/x002_gfsv17_issue_1219/x002_gfsv17_issue_1219.xml
```

The resulting Rocoto workflow is as follows.

```
       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
202103231200             gfsstage_ic                    51952235              QUEUED                   -         0           0.0
202103231200                 gfsfcst                           -                   -                   -         -             -
202103231200            gfspost_f000                    51952236              QUEUED                   -         0           0.0
202103231200            gfspost_f006                    51952237              QUEUED                   -         0           0.0
202103231200            gfspost_f012                    51952238              QUEUED                   -         0           0.0
202103231200                 gfsvrfy                           -                   -                   -         -             -
202103231200             gfsmetpg2g1                           -                   -                   -         -             -
202103231200             gfsmetpg2o1                           -                   -                   -         -             -
202103231200             gfsmetppcp1                           -                   -                   -         -             -
202103231200               gfsgempak                           -                   -                   -         -             -
202103231200           gfsgempakmeta                           -                   -                   -         -             -
202103231200       gfsgempakmetancdc                           -                   -                   -         -             -
202103231200    gfsgempakncdcupapgif                           -                   -                   -         -             -
202103231200      gfsgempakpgrb2spec                           -                   -                   -         -             -
202103231200                 gfsarch                           -                   -                   -         -             -
202103231200              gfscleanup                           -                   -                   -         -             -

```

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: workflow/applications/gfs_cycled.py [hunk 1]
```diff
@@ -89,9 +89,6 @@ def _get_app_configs(self):
         if self.do_awips:
             configs += ['awips']
 
-        if self.do_npoess:
-            configs += ['npoess']
-
```
Comment (aerorahul) [inline workflow/applications/gfs_cycled.py:94]: this needs to be added to the list of configs if `do_gempak` is yes on line 84.

File: workflow/applications/gfs_cycled.py [hunk 2]
```diff
@@ -230,13 +231,14 @@ def get_task_names(self):
 
         if self.do_gempak:
             gfs_tasks += ['gempak']
+            gfs_tasks += ['gempakmeta']
+            gfs_tasks += ['gempakncdcupapgif']
+            gfs_tasks += ['gempakpgrb2spec']
+            gfs_tasks += ['npoess_pgrb2_0p5deg']
```
Comment (aerorahul) [inline workflow/applications/gfs_cycled.py]:
```suggestion
            gfs_tasks += ['gempak', 'gempakmeta', 'gempakncdcupagif', 'npoess_pgrb2_0p5deg', 'gempakpgrb2spec']
```
gempakpgrb2spec depends on npoess_pgrb2_0p5deg so it should come first.  order matters in the xml

File: workflow/applications/gfs_forecast_only.py [hunk 3]
```diff
@@ -113,6 +113,9 @@ def get_task_names(self):
 
             if self.do_gempak:
                 tasks += ['gempak']
+                tasks += ['gempakmeta']
+                tasks += ['gempakncdcupapgif']
+                tasks += ['gempakpgrb2spec']
```
Comment (aerorahul) [inline workflow/applications/gfs_forecast_only.py]:
```suggestion
                tasks += ['gempak', 'gempakmeta', 'gempakncdcupagif', 'gempakpgrb2spec']
```

File: workflow/applications/gfs_cycled.py [hunk 4]
```diff
@@ -187,6 +184,10 @@ def get_task_names(self):
         if self.do_vminmon:
             gdas_tasks += ['vminmon']
 
+        if self.do_gempak:
+            gdas_tasks += ['gempak']
+            gdas_tasks += ['gempakmetancdc']
```
Comment (aerorahul) [inline workflow/applications/gfs_cycled.py]:
```suggestion
            gdas_tasks += ['gempak', 'gempakmetancdc']
```

File: workflow/rocoto/gfs_tasks.py [hunk 5]
```diff
@@ -935,15 +935,58 @@ def gempak(self):
 
         return task
 
-    def npoess(self):
+    def gempakmeta(self):
+        deps = []
+        dep_dict = {'type': 'metatask', 'name': f'{self.cdump}post'}
```
Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]: the dependency on `f{self.cdump}post` is no longer correct. This metatask does not exist after #2110

File: workflow/rocoto/gfs_tasks.py [hunk 6]
```diff
@@ -935,15 +935,58 @@ def gempak(self):
...
+        dependencies = rocoto.create_dependency(dep=deps)
+
+        resources = self.get_resource('gempak')
+        task = create_wf_task('gempakmeta', resources, cdump=self.cdump, envar=self.envars, dependency=dependencies)
+
+        return task
+
+    def gempakmetancdc(self):
+        deps = []
+        dep_dict = {'type': 'metatask', 'name': f'{self.cdump}post'}
```
Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]: same comment as above on the dependency.

File: workflow/rocoto/gfs_tasks.py [hunk 7]
```diff
@@ -935,15 +935,58 @@ def gempak(self):
...
+        dependencies = rocoto.create_dependency(dep=deps)
+
+        resources = self.get_resource('gempak')
+        task = create_wf_task('gempakmetancdc', resources, cdump=self.cdump, envar=self.envars, dependency=dependencies)
+
+        return task
+
+    def gempakncdcupapgif(self):
+        deps = []
+        dep_dict = {'type': 'metatask', 'name': f'{self.cdump}post'}
```
Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]: same comment as above on the dependency.

**Review by aerorahul — CHANGES_REQUESTED** [review]
mostly looks good now.
1. some consistency suggestions
2. dependencies for the new gempak jobs need to be corrected after #2110

File: workflow/applications/gfs_cycled.py (hunk 1 above)
Comment (HenryRWinterbottom) [reply workflow/applications/gfs_cycled.py:94]: Done.

File: workflow/rocoto/gfs_tasks.py [hunk 8]
```diff
@@ -935,15 +935,58 @@ def gempak(self):
...
+        dependencies = rocoto.create_dependency(dep=deps)
+
+        resources = self.get_resource('gempak')
+        task = create_wf_task('gempakncdcupapgif', resources, cdump=self.cdump, envar=self.envars, dependency=dependencies)
+
+        return task
+
+    def gempakpgrb2spec(self):
+        deps = []
+        dep_dict = {'type': 'metatask', 'name': f'{self.cdump}npoess_pgrb2_0p5deg'}
```
Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]: `npoes_pgrb2_0p5deg` is not a metatask.  It is a task.

Comment (aerorahul) [inline workflow/rocoto/gfs_tasks.py]:
```suggestion
        dep_dict = {'type': 'task', 'name': f'{self.cdump}npoess_pgrb2_0p5deg'}
```

File: workflow/applications/gfs_cycled.py [hunk 9]
```diff
@@ -229,14 +229,11 @@ def get_task_names(self):
             gfs_tasks += ['postsnd']
 
         if self.do_gempak:
-            gfs_tasks += ['gempak']
+            gfs_tasks += ['gempak', 'gempakmeta', 'gempakncdcupagif', 'gempakpgrb2spec', 'npoes_pgrb2_0p5deg']
```
Comment (aerorahul) [inline workflow/applications/gfs_cycled.py]:
```suggestion
            gfs_tasks += ['gempak', 'gempakmeta', 'gempakncdcupagif', 'npoess_pgrb2_0p5deg', 'gempakpgrb2spec']
```

File: workflow/rocoto/gfs_tasks.py (hunk 5 above)
Comment (HenryRWinterbottom) [reply workflow/rocoto/gfs_tasks.py]: @aerorahul Is the new dependency `upp`, `atmos_products`, or both?

File: workflow/applications/gfs_cycled.py (hunk 9 above)
Comment (aerorahul) [reply workflow/applications/gfs_cycled.py]: Typo and the task order matters.
`gempakpgrb2spec` depends on `npoess_pgrb2_0p5deg`, so it should come first.

File: workflow/applications/gfs_cycled.py [hunk 10]
```diff
@@ -82,16 +82,14 @@ def _get_app_configs(self):
 
         if self.do_gempak:
             configs += ['gempak']
+            configs += ['npoess']
```
Comment (aerorahul) [inline workflow/applications/gfs_cycled.py]:
```suggestion
            configs += ['gempak', 'npoess']
```

File: workflow/rocoto/gfs_tasks.py (hunk 5 above)
Comment (aerorahul) [reply workflow/rocoto/gfs_tasks.py]: `atmprod` I think.

File: workflow/applications/gfs_cycled.py (hunk 9 above)
Comment (HenryRWinterbottom) [reply workflow/applications/gfs_cycled.py]: Done.

File: workflow/rocoto/gfs_tasks.py (hunk 6 above)
Comment (HenryRWinterbottom) [reply workflow/rocoto/gfs_tasks.py]: Done.

File: workflow/rocoto/gfs_tasks.py (hunk 5 above)
Comment (HenryRWinterbottom) [reply workflow/rocoto/gfs_tasks.py]: Done.

File: workflow/rocoto/gfs_tasks.py (hunk 7 above)
Comment (HenryRWinterbottom) [reply workflow/rocoto/gfs_tasks.py]: Done.

File: workflow/rocoto/gfs_tasks.py (hunk 8 above)
Comment (HenryRWinterbottom) [reply workflow/rocoto/gfs_tasks.py]: Done.

File: workflow/rocoto/gfs_tasks.py (hunk 5 above)
Comment (WalterKolczynski-NOAA) [reply workflow/rocoto/gfs_tasks.py]: Doesn't look like it. Did you forget to push?

File: workflow/rocoto/gfs_tasks.py (hunk 8 above)
Comment (WalterKolczynski-NOAA) [reply workflow/rocoto/gfs_tasks.py]: This still haven't been updated, or the fix got reverted.

**Review by aerorahul — APPROVED** [review]
looks good now.

**Pull request merged**